cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/cluster/Diffs.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.controller.cluster.operations.cluster;
+
+import io.fabric8.kubernetes.api.model.Container;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.extensions.StatefulSet;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiPredicate;
+
+import static io.strimzi.controller.cluster.resources.AbstractCluster.containerEnvVars;
+
+class Diffs {
```

Comment:
I think I need to call this out for particular attention, because my reasoning might be flawed, or I might have missed details... Previously we were using the `diff()` method on the subclasses of `AbstractCluster` to figure out a few things (scaling, rolling update, whether there were "differences"). We no longer need a number of these:

* The scaling I pushed the logic down into the scalable resources (SS and Dep) so that it's a noop if the resource doesn't exist yet (the create case), or if the thing has a larger/smaller scale than we're scaling up/down to.
* The differences, as I mentioned, we basically just always patch. Now we're being much stricter with what those resources look like because we always patch *based on the state of the cluster CM*.
* As I mentioned in the summary, we still care about when a rolling update is necessary. I've *reimplmented* that logic for Kafka and ZK, but I might have screwed it up, and I haven't yet done it for Connect. 

So please review `Diffs`, `KafkaSetOperations` and `ZookeeperSetOperations` carefully in these respects.

Reply:
> As I mentioned in the summary, we still care about when a rolling update is necessary. I've reimplmented that logic for Kafka and ZK, but I might have screwed it up, and I haven't yet done it for Connect.

I might be wrong ... but I think that for Connect we do not do rolling update ... we just leave it for Kubernetes to deal with it. This is becasue it basically stateless deployment and there was so far no special need to take care of this by us.